docs/design-documents/platform/Statistics/NetworkStatistics.md

```diff
@@ -0,0 +1,186 @@
+}mbed_stats_type_t;
+```
+1. Statistic Info Structure - Structure used to get statistics data as binary blob from the StatsList class.
+
+```
+typedef struct  {
+    mbed_stats_type_t stats_type;
+    uint32_t buf_size;
+    void *buf;
+}stats_info_t;
+```
+
+#### API's
+
+1. static int get_each(stats_info_t *stats, int count);
+API to fetch all the statistics data available in the linked list. User can specify how many elements from linked list should be fetched in `count` parameter.
+
+2. static int get_each(stats_info_t *stats, mbed_stats_type_t type, int count);
+API to fetch specific `type` of statistic data available in the linked list. User can specify how many elements from linked list should be fetched in `count` parameter.
+
+## Ethernet Statistics
+
+EMAC interface class and LWIP stack together provided all information required for ethernet stattistics. `EMACInterfaceStats` and `LWIPStackStats` classes provide mechanism to log all statistics information.
+API's (`log_xxx`) provided by `EMACInterfaceStats` should be used by `EMACInterface` source to log interface related information such as IP address, netmask, etc.
+API's (`log_xxx`) provided by `LWIPStackStats` should be used by `LWIP` source to log network stack information such as tcp and udp sockets open, data packets transmitted, data packets received, etc.
+
+#### API Usage
+```C
+
+void get_interface_stats ()
+{
+    stats_info_t stats;
+    nw_emac_info_t *buf;
+    StatsList::get_each(&stats, STATS_NW_INTERFACE_EMAC, 1);
+    buf = (nw_emac_info_t *)stats.buf;
+    printf("IP %s \t Mac %s\n", buf->ip, buf->mac);
+}
+
+void get_stack_stats()
+{
+    stats_info_t stats;
+    nw_stack_stats_t *buf;
+    StatsList::get_each(&stats, STATS_NW_STACK_LWIP, 1);
+    buf = (nw_stack_stats_t *)stats.buf;
+    printf("Stack addr = %p\n", (void *)buf->stack_addr);
+    printf("TCP sockets open %d UDP sockets open %d \n", buf->tcp.sock_open, buf->udp.sock_open);
+    printf("TCP xmit count %d UDP xmit count %d \n", buf->tcp.pkt_xmit, buf->udp.pkt_xmit);
+}
+
+```
+
+#### Data structures & Member functions
+
+1. EMAC interface can provided IP, netmask, connection status, underlying hardware interface specifics, etc. All this information is added to `nw_emac_info_t` structure.
+```
+typedef struct {
+    uint32_t stack_addr;
+    uint8_t ip[NSAPI_IPv6_SIZE];
+    uint8_t gateway[NSAPI_IPv4_SIZE];
+    uint8_t netmask[NSAPI_IPv4_SIZE];
+    uint8_t mac[NSAPI_MAC_SIZE];
+    uint8_t connect_status;
+    bool dhcp;
+}nw_emac_info_t;
+```
+
+1. Network stack information related to TCP and UDP sockets is collected as part of `nw_stack_stats_t` structure.
+```
+typedef struct {
+    uint32_t pkt_xmit;         /* Transmitted packets. */
+    uint32_t pkt_recv;         /* Received packets. */
+    uint16_t sock_open;        /* Number of sockets opened */
+}stack_stats_t;
+
+typedef struct {
+    uint32_t stack_addr;
+    stack_stats_t udp;
+    stack_stats_t tcp;
+}nw_stack_stats_t;
```

Comment:
I don't see the purpose of having UDP and TCP stats separately. There might be other protocols as well (NB-IoT, LoRA). We should aim to be more generic.

It is already known in the design phase of the application which protocol they use.

Reply:
Please suggest what all should be added/removed from this structure. We can have common count for udp/tcp sockets if that is more meaningful